Gantt state docs (#1090)
telerik/blazor-docs: components/gantt/state.md

---
title: State
page_title: Gantt - State
description: Save, load, change the state of the Gantt for Blazor - sorting, filtering and more.
slug: gantt-state
tags: telerik,blazor,gantt,state,save,load,layout,get,set,change
published: True
position: 15
---

> # Gantt State

Reviewer: It's good to link to the GanttState API somewhere or even at a few different places. The URL should be `/blazor-ui/api/Telerik.Blazor.Components.GanttState-1`

> The Gantt lets you save, load and change its current state through code. The state management includes all the user-configurable elements of the Gantt - such as sorting, filtering, edited items, TreeList width, column size and order.

Reviewer: ```suggestion
The Gantt lets you save, load and change its current state through code. The state management includes all the user-configurable properties of the Gantt - such as sorting, filtering, edited items, TreeList width, column size and order.
```

> You can see this feature in the [Live Demo: Gantt State](https://demos.telerik.com/blazor-ui/Gantt/persist-state).

Reviewer: ```suggestion
See the feature in action in the [Live Demo: Gantt State](https://demos.telerik.com/blazor-ui/gantt/persist-state).
```

This article contains the following sections:

<!-- Start Document Outline -->

* [Basics](#basics)
	* [Events](#events)
	* [Methods](#methods)
* [Information in the Gantt State](#information-in-the-gantt-state)
* [Examples](#examples)
	* [Save and Load Gantt State from Browser LocalStorage](#save-and-load-gantt-state-from-browser-localstorage)
	* [Set Gantt Options Through State](#set-gantt-options-through-state)
	* [Set Default (Initial) State](#set-default-initial-state)
	* [Get and Override User Action That Changes The Gantt](#get-and-override-user-action-that-changes-the-gantt)
	* [Initiate Editing or Inserting of an Item](#initiate-editing-or-inserting-of-an-item)
	* [Get Current Columns Visibility, Order, Field](#get-current-columns-visibility-order-field)

<!-- End Document Outline -->

## Basics

> The Gantt state is a generic class whose type is determined by the type of the model you use for the Gantt. It contains fields that correspond to the Gantt behaviors which you can use to save, load and modify the Gantt state.

Reviewer: ```suggestion
`GanttState<TItem>` is a generic class whose type is determined by the type of the model you use for the Gantt. The state object contains properties that correspond to the Gantt behaviors. You can modify, save and load the Gantt state.
```

Fields that pertain to model data (such as edited item, inserted item, expanded items) are also typed according to the Gantt model. If you restore such data, make sure to implement appropriate comparison checks - by default the `.Equals `check for a class (model) is a reference check and the reference from the storage is unlikely to match the reference from the Gantt `Data`. Thus, you may want to override the `.Equals` method of the model you use so it compares by an ID, for example, or otherwise (in the app logic) re-populate the models in the state object with the new model references from the Gantt data source.

The Gantt offers two events and two methods to allow flexible operations over its state:

* [Events](#events)

> 

Reviewer: ```suggestion
```

* [Methods](#methods)

### Events

> The `OnStateInit` and `OnStateChanged` events are raised by the Gantt so you can have an easy to use hook for loading and saving state, respectively.

Reviewer: ```suggestion
The `OnStateInit` and `OnStateChanged` events are raised by the Gantt so you can have easy-to-use hooks for loading and saving state, respectively.
```

* `OnStateInit` fires when the Gantt is initializing and you can provide the state you load from your storage to the `State` field of its event arguments.

> * `OnStateChanged` fires when the user makes a change to the Gantt state (such as sorting, filtering, editing, expanding parents and so on). The `State` field of the event argument provides the current Gantt state so you can store it. The `PropertyName` field of the event arguments indicates what is the aspect that changed.

Reviewer: ```suggestion
* `OnStateChanged` fires when the user makes a change to the Gantt state (such as sorting, filtering, editing, expanding parents and so on). The `State` field of the event argument provides the current Gantt state so you can store it. The `PropertyName` field of the event arguments indicates what changed.
```

If you want to make changes on the current Gantt state, first get it from the Gantt through the `GetState` method, then apply the modifications on the object you got and pass it to `SetStateAsync`. This will allow you preserving the current component options such as TreeListWidth, Timeline View, sorts, filters etc. 

If you want to put the Gantt in a certain configuration without preserving the old one, create a `new GanttState<T>()` and apply the settings there, then pass it to `SetStateAsync`. Thus, any previously applied settings will be lost and the component options will be reset to their default state if not explicitly set through the `GanttState`.

To reset the Gantt state, call `SetStateAsync(null)`.

You should avoid calling `SetStateAsync` in the Gantt [CRUD methods]({%slug gantt-tree-editing%}#events) (such as `OnUpdate`, `OnEdit`, `OnCreate`). Doing so may lead to unexpected results because the Gantt has more logic to execute after the event.

> ## Information in the Gantt State

Reviewer: We should mention the GanttState property names that relate to the list items, or link to the GanttState API. This will make the documentation more actionable.

    }
}
````

<!-- ### Save and Load Gantt State in a WebAssembly application
The [knowledge base article for saving the Gantt state in a WASM application] explains two ways of storing the `Gantt` state - through a custom controller and a custom service that calls the browser's LocalStorage. -->

### Set Gantt Options Through State

The Gantt state allows you to control the behavior of the Gantt programmatically - you can, for example, set sorts, filteres, expand hierarhical items.

>tip The individual tabs below show how you can use the state to programmatically set the Gantt filtering, sorting and other features.

>tip If you want to set an initial state to the Gantt, use a similar snippet, but in the `OnStateInit` event.

<div class="skip-repl"></div>
````Sorting
@[template](/_contentTemplates/gantt/state.md#set-sort-from-code)
````
````FilterRow
@[template](/_contentTemplates/gantt/state.md#filter-row-from-code)
````
````FilterMenu
@[template](/_contentTemplates/gantt/state.md#filter-menu-from-code)
````
````Hierarchy
@[template](/_contentTemplates/gantt/state.md#expand-hierarchy-from-code)
````

@[template](/_contentTemplates/gantt/state.md#filter-menu-default-filters)

> ### Set Default (Initial) State

Reviewer: This example is one of the fundamental ones and should be first. Loading and Saving the state are based on this one, so that example should come after this one.

                        End = new DateTime(2021, 7, 6 + i + j),
                        PercentComplete = Math.Round(random.NextDouble(), 2)
                    });

                LastId++;
            }
        }

        base.OnInitialized();
    }

    private void UpdateItem(GanttUpdateEventArgs args)
    {
        var item = args.Item as GanttTask;

        var foundItem = Data.FirstOrDefault(i => i.Id.Equals(item.Id));

        if (foundItem != null)
        {
            foundItem.Title = item.Title;
            foundItem.Start = item.Start;
            foundItem.End = item.End;
            foundItem.PercentComplete = item.PercentComplete;
        }
    }

    private void DeleteItem(GanttDeleteEventArgs args)
    {
        var item = Data.FirstOrDefault(i => i.Id.Equals((args.Item as GanttTask).Id));

        RemoveChildRecursive(item);
    }

    private void RemoveChildRecursive(GanttTask item)
    {
        var children = Data.Where(i => item.Id.Equals(i.ParentId)).ToList();

        foreach (var child in children)
        {
            RemoveChildRecursive(child);
        }

        Data.Remove(item);
    }
}
````

> ### Initiate Editing or Inserting of an Item

Reviewer: Customers often ask how to EXIT edit mode programmatically, so we need such a button in this example too.

            if (children.Any())
            {
                parent.PercentComplete = children.Average(i => i.PercentComplete);

                CalculateParentPercentRecursive(parent);
            }
        }
    }

    private void CalculateParentRangeRecursive(GanttTask item)
    {
        if (item.ParentId != null)
        {
            var parent = GetParent(item);

            var children = GetChildren(parent);

            if (children.Any())
            {
                parent.Start = children.Min(i => i.Start);
                parent.End = children.Max(i => i.End);

                CalculateParentRangeRecursive(parent);
            }
        }
    }

    private GanttTask GetParent(GanttTask item)
    {
        return Data.FirstOrDefault(i => i.Id.Equals(item.ParentId));
    }

    private IEnumerable<GanttTask> GetChildren(GanttTask item)
    {
        return Data.Where(i => item.Id.Equals(i.ParentId));
    }
}
````

### Get Current Columns Visibility, Order, Field

The `ColumnStates` property of the `GanttState` object provides you with information about the current state of the Gantt columns. It contains the following properties:

> Field | Type | Description

Reviewer: The table is missing vertical borders at both ends. If we agree to use this approach, we should be consistent. (I personally am reluctant about this, as it hinders readability)

---------|----------|---------
 `Index` | `int` | the current index of the column based on the position the user chose
 `Id` | `string` | the Id of the column if it is set
 `Field` | `string` | the field of the column
 `Visible` | `bool?` | whether the column is hidden or not
 `Width` | `string` | the width of the column if it is set

> By looping over the `ColumnStates` collection you can know what the user sees. By default, the order of the columns in the state collection will remain the same but their `Index` value will be changed to indicate their position. You can, for example, sort by the index and filter by the visibility of the columns to get the approximate view the user sees.

Reviewer: ```suggestion
By looping over the `ColumnStates` collection you can know what the user sees. By default, the order of the columns in the state collection will remain the same but their `Index` value will change to indicate their position. You can, for example, sort by the index and filter by the visibility of the columns to get the approximate view the user sees.
```
